osquery/tables/system/windows/prefetch.cpp

```diff
@@ -0,0 +1,370 @@
+  DWORD NumberOfVolumes;
+  DWORD VolumesInformationSize;
+  union {
+    struct {
+      DWORD Reserved1[2];
+      FILETIME LastRunTime;
+      DWORD Reserved2[4];
+      DWORD RunCount;
+    } v23;
+    struct {
+      DWORD Reserved1[2];
+      FILETIME LastRunTime;
+      FILETIME OtherRunTimes[7];
+      DWORD Reserved2[4];
+      DWORD RunCount;
+    } v26;
+    struct {
+      DWORD Reserved1[2];
+      FILETIME LastRunTime;
+      FILETIME OtherRunTimes[7];
+      DWORD Reserved2[2];
+      DWORD RunCount;
+    } v30v2;
+  } ext;
+} PREFETCH_FILE_INFORMATION, *PPREFETCH_FILE_INFORMATION;
+
+typedef struct _PREFETCH_VOLUME_INFORMATION {
+  DWORD VolumePathOffset;
+  DWORD VolumeDevicePathNumberOfCharacters;
+  FILETIME VolumeCreationTime;
+  DWORD VolumeSerialNumber;
+  DWORD FileReferencesOffset;
+  DWORD FileReferencesDataSize;
+  DWORD DirectoryStringsOffset;
+  DWORD NumberOfDirectoryStrings;
+  DWORD Reserved1;
+} PREFETCH_VOLUME_INFORMATION, *PPREFETCH_VOLUME_INFORMATION;
+
+typedef struct _DIRECTORY_STRING {
+  USHORT Size;
+  WCHAR Directory[1]; // arbitrary size
+} DIRECTORY_STRING, *PDIRECTORY_STRING;
+#pragma pack(pop)
+
+} // namespace
+
+PrefetchHeader parseHeader(const PREFETCH_FILE_HEADER* header) {
+  PrefetchHeader result;
+  if (header->FileName[(sizeof(header->FileName) / 2) - 1] == '\0') {
+    result.filename = wstringToString(header->FileName);
+  }
+  result.prefetch_hash = (boost::format("%x") % header->Hash).str();
+  result.file_size = header->FileSize;
+  return result;
+}
+
+PrefetchFileInfo parseFileInfo(
+    const std::vector<UCHAR>& data,
+    const PREFETCH_FILE_INFORMATION* prefetch_file_info,
+    DWORD version) {
+  PrefetchFileInfo result;
+
+  FILETIME last_run_time;
+  if (version == 23) {
+    last_run_time = prefetch_file_info->ext.v23.LastRunTime;
+    result.run_count = prefetch_file_info->ext.v23.RunCount;
+  } else if (version == 26) {
+    last_run_time = prefetch_file_info->ext.v26.LastRunTime;
+    result.run_count = prefetch_file_info->ext.v26.RunCount;
+  } else if (version == 30) {
+    last_run_time = prefetch_file_info->ext.v30v2.LastRunTime;
+    result.run_count = prefetch_file_info->ext.v30v2.RunCount;
+  }
+  result.last_run_time = filetimeToUnixtime(last_run_time);
+
+  // Size is given in bytes.
+  const auto size = prefetch_file_info->FileNameStringsSize;
+  const auto offset = prefetch_file_info->FileNameStringsOffset;
+  if (offset > data.size()) {
+    // Unexpected offset.
+    return result;
+  }
+
+  size_t total_length{0};
+  std::vector<std::string> filenames;
+  auto next = (PWCHAR)(&data[0] + offset);
+  while (*next != '\0') {
+    auto length = wcsnlen_s(next, (size - total_length) / 2);
+    if (length == 0 || length == (size - total_length) / 2) {
+      // A null wide character was not found.
+      break;
+    }
+
+    auto filename = wstringToString(next);
+    filenames.emplace_back(std::move(filename));
+    total_length += (length + 1) * 2;
```

Comment:
recommend using `* sizeof(WCHAR)` vs `* 2`

Reply:
switched to sizeof for all * 2